src/uu/tail/src/chunks.rs

```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    /// `chunk.lines` then a new empty `BytesChunk` is returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `chunk`: The chunk to create a new `BytesChunk` chunk from
+    /// * `offset`: Start to copy the old chunk's buffer from this position. May not be larger
+    ///             than `chunk.bytes`.
+    ///
+    /// returns: BytesChunk
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let mut chunk = BytesChunk::new();
+    /// chunk.buffer[1] = 1;
+    /// chunk.bytes = 2;
+    /// let new_chunk = BytesChunk::from_chunk(&chunk, 0);
+    /// assert_eq!(2, new_chunk.get_buffer().len());
+    /// assert_eq!(&[0, 1], new_chunk.get_buffer());
+    ///
+    /// let new_chunk = BytesChunk::from_chunk(&chunk, 1);
+    /// assert_eq!(1, new_chunk.get_buffer().len());
+    /// assert_eq!(&[1], new_chunk.get_buffer());
+    /// ```
+    fn from_chunk(chunk: &Self, offset: usize) -> Self {
+        if offset >= chunk.bytes {
+            return Self::new();
+        }
+
+        let mut buffer: ChunkBuffer = [0; BUFFER_SIZE];
+        let slice = chunk.get_buffer_with(offset);
+        buffer[..slice.len()].copy_from_slice(slice);
+        Self {
+            buffer,
+            bytes: chunk.bytes - offset,
+        }
+    }
+
+    /// Receive the internal buffer safely, so it returns a slice only containing as many bytes as
+    /// large the `self.bytes` value is.
+    ///
+    /// returns: a slice containing the bytes of the internal buffer from `[0..self.bytes]`
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let mut chunk = BytesChunk::new();
+    /// chunk.bytes = 1;
+    /// assert_eq!(&[0], chunk.get_buffer());
+    /// ```
+    pub(crate) fn get_buffer(&self) -> &[u8] {
+        &self.buffer[..self.bytes]
+    }
+
+    /// Like [`BytesChunk::get_buffer`], but returning a slice from `[offset.self.bytes]`.
+    ///
+    /// returns: a slice containing the bytes of the internal buffer from `[offset..self.bytes]`
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let mut chunk = BytesChunk::new();
+    /// chunk.bytes = 2;
+    /// assert_eq!(&[0], chunk.get_buffer_with(1));
+    /// ```
+    pub(crate) fn get_buffer_with(&self, offset: usize) -> &[u8] {
+        &self.buffer[offset..self.bytes]
+    }
+
+    pub(crate) fn has_data(&self) -> bool {
+        self.bytes > 0
+    }
+
+    /// Fills this [`buffer`](Self::buffer) with maximal [`BUFFER_SIZE`] number of bytes, draining
+    /// the reader by that number of bytes. If EOF is reached (so 0 bytes are read), then returns
+    /// [`UResult<None>`] or else the result with [`Some(bytes)`] where bytes is the number of bytes
+    /// read from the source.
+    ///
+    /// # Arguments
+    ///
+    /// * `filehandle`: A [`BufReader`], where the inner must implement [`Read`]
+    ///
+    /// returns: `Result<Option<usize>, Box<dyn UError, Global>>`
```

Comment:
The documentation for the argument and return type do not add much over the type signature here. I think we can remove them. The same goes for some other functions.